In [48]:
pip install --prefix /usr/local joblib

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


**Objective:Making a model based on logistic regression and predicting values based on it.The dataset used for it is "Credit card frad detection dataset"**

---




In [50]:
#importing all the necessary libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import joblib


In [51]:
#loading dataset 
credit_card_data=pd.read_csv('/content/creditcard.csv')
credit_card_data


Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [7]:
#dataset informations
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156612 entries, 0 to 156611
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    156612 non-null  float64
 1   V1      156612 non-null  float64
 2   V2      156611 non-null  float64
 3   V3      156611 non-null  float64
 4   V4      156611 non-null  float64
 5   V5      156611 non-null  float64
 6   V6      156611 non-null  float64
 7   V7      156611 non-null  float64
 8   V8      156611 non-null  float64
 9   V9      156611 non-null  float64
 10  V10     156611 non-null  float64
 11  V11     156611 non-null  float64
 12  V12     156611 non-null  float64
 13  V13     156611 non-null  float64
 14  V14     156611 non-null  float64
 15  V15     156611 non-null  float64
 16  V16     156611 non-null  float64
 17  V17     156611 non-null  float64
 18  V18     156611 non-null  float64
 19  V19     156611 non-null  float64
 20  V20     156611 non-null  float64
 21  V21     15

In [6]:
#checking the number of missing values in each column
credit_card_data.isnull().sum()

Time      0
V1        0
V2        1
V3        1
V4        1
V5        1
V6        1
V7        1
V8        1
V9        1
V10       1
V11       1
V12       1
V13       1
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [10]:
#in the dataset class label for legit or legal transaction is marked as '0' and illegal is marked as '1'
credit_card_data['Class'].value_counts()

0.0    187892
1.0       368
Name: Class, dtype: int64

**So,we can see that dataset is highly imbalanced as number of entries for legal transaction is 187892 and that of illegal is 368 so we need to make a sample of uniform dataset of both entries almost equal** 

In [11]:
#seperating the dataset for analysis
legit=credit_card_data[credit_card_data.Class==0]
fraud=credit_card_data[credit_card_data.Class==1]

In [12]:
#printing both legit and fraud
print(legit.shape)
print(fraud.shape)

(187892, 31)
(368, 31)


In [13]:
#statistical measures of the data
legit.Amount.describe()

count    187892.000000
mean         88.949661
std         247.532742
min           0.000000
25%           5.960000
50%          22.750000
75%          79.000000
max       19656.530000
Name: Amount, dtype: float64

In [14]:
fraud.Amount.describe()

count     368.000000
mean      115.659891
std       248.444604
min         0.000000
25%         1.000000
50%        11.395000
75%       104.007500
max      2125.870000
Name: Amount, dtype: float64

In [15]:
#compare both the values of fraud and legit
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    66928.356710 -0.115473  0.005839  0.398688  0.079599 -0.127383   
1.0    58116.711957 -5.926741  4.330987 -7.884388  4.836685 -4.286202   

            V6        V7        V8        V9  ...       V20       V21  \
Class                                         ...                       
0.0    0.05489 -0.048754  0.021358  0.018046  ...  0.023277 -0.020226   
1.0   -1.42407 -6.893932  0.739547 -2.838131  ...  0.404833  0.790022   

           V22       V23       V24       V25       V26       V27       V28  \
Class                                                                        
0.0   -0.06081 -0.018125  0.006229  0.072192  0.008418  0.001399  0.002064   
1.0   -0.02306 -0.043940 -0.071941  0.080214  0.040915  0.177371  0.050569   

           Amount  
Class              
0.0     88.949661  
1.0    115.659891  

[2 rows x 30 columns]

**So the data is highly imbalanced so we are taking random values for 368 legit dataset as fraud contains 368**

In [16]:
#sample space for legit dataset
legit_sample=legit.sample(n=368)

**Now concatenating both the datasets of fraud and legit_sample**

In [17]:
#Concatenating both dataset
new_dataset=pd.concat([legit_sample,fraud],axis=0)
new_dataset

Time        V1        V2        V3        V4        V5        V6  \
168573  119288.0  0.124319  0.619917 -1.019457 -2.723670  1.155949 -1.522562   
168292  119155.0 -1.232672  1.038254  0.048484 -2.419244 -0.140232  0.059805   
173752  121628.0 -2.006809  0.908377  0.163097 -0.913962 -0.914933 -0.372983   
117053   74524.0  0.908828 -1.404905  1.314052  0.959853 -1.777457  0.705762   
159299  112396.0  2.025331 -0.126615 -1.419791  0.079005  0.209970 -0.558672   
...          ...       ...       ...       ...       ...       ...       ...   
178208  123525.0 -5.904921  4.439911 -8.631802  7.788684 -4.989580 -1.200144   
181966  125200.0 -0.769172  1.342212 -2.171454 -0.151513 -0.648374 -0.973504   
182992  125612.0  1.889618  1.073099 -1.678018  4.173268  1.015516 -0.009389   
183106  125658.0  0.224414  2.994499 -3.432458  3.986519  3.760233  0.165640   
184379  126219.0 -1.141559  1.927650 -3.905356 -0.073943 -0.044858 -1.756999   

              V7        V8        V9  ...       V21       V22       V23  \
168573  1.594983 -0.486019  0.368096  ...  0.256683  0.850103 -0.192944   
168292  0.981257 -2.277936 -1.359164  ...  1.629433 -0.237162 -0.184905   
173752  0.108699  0.181346  1.146667  ... -0.070509 -0.171970  0.104005   
117053 -1.157607  0.260878  0.600663  ... -0.337695 -0.480577 -0.215476   
159299  0.010150 -0.092658  0.573483  ...  0.319005  1.009354  0.021405   
...          ...       ...       ...  ...       ...       ...       ...   
178208 -7.674060  4.125761 -5.315778  ...  1.775030  1.266441 -0.199410   
181966 -1.706658  0.313745 -1.982302  ... -0.036122 -0.753591 -0.047113   
182992 -0.079706  0.064071 -0.714517  ...  0.203728  0.733796 -0.036560   
183106  1.099378 -0.654557 -2.435416  ...  0.491337 -0.984223 -0.421979   
184379 -1.217416  0.364563 -2.770148  ...  0.102081 -0.531496 -0.328741   

             V24       V25       V26       V27       V28  Amount  Class  
168573  0.664661 -0.338198 -0.270050  0.174619  0.215630   13.60    0.0  
168292 -0.470986  0.068671 -0.371505 -0.590804 -0.168054  133.84    0.0  
173752  0.050061 -0.855265 -0.531930 -1.459421 -0.335494   62.92    0.0  
117053 -0.125348  0.407634 -0.249811  0.092181  0.068683  187.19    0.0  
159299  0.712812  0.280713 -0.467959 -0.000705 -0.058316    1.00    0.0  
...          ...       ...       ...       ...       ...     ...    ...  
178208  0.014960 -0.000019  0.544210 -1.167566 -1.206354   45.51    1.0  
181966  0.358493 -0.287407  0.476506  0.250531  0.250987   40.00    1.0  
182992  0.334306  0.147171  0.279556  0.031669  0.035883    3.22    1.0  
183106 -1.048058  0.726412  0.268625  0.283689  0.419102    0.00    1.0  
184379  0.393100  0.568435  0.786605 -0.146102  0.076211   25.00    1.0  

[736 rows x 31 columns]

In [18]:
new_dataset['Class'].value_counts()

0.0    368
1.0    368
Name: Class, dtype: int64

**Thus now the dataset is a good fit for testing and training**

In [20]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    66313.788043 -0.134189 -0.095668  0.371460  0.051690 -0.194387   
1.0    58116.711957 -5.926741  4.330987 -7.884388  4.836685 -4.286202   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0   -0.000799 -0.045916  0.038212  0.044558  ...  0.029031 -0.044945   
1.0   -1.424070 -6.893932  0.739547 -2.838131  ...  0.404833  0.790022   

           V22       V23       V24       V25       V26       V27       V28  \
Class                                                                        
0.0   -0.04938 -0.000431 -0.016758  0.104893  0.016293 -0.001118  0.013289   
1.0   -0.02306 -0.043940 -0.071941  0.080214  0.040915  0.177371  0.050569   

           Amount  
Class              
0.0    101.668886  
1.0    115.659891  

[2 rows x 30 columns]

**Splitting the dataset into "Features and "Targets"**

In [23]:
#Splitting the dataset into "Features and "Targets"
X=new_dataset.drop(columns='Class',axis=1)
Y=new_dataset['Class']

In [24]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
168573  119288.0  0.124319  0.619917 -1.019457 -2.723670  1.155949 -1.522562   
168292  119155.0 -1.232672  1.038254  0.048484 -2.419244 -0.140232  0.059805   
173752  121628.0 -2.006809  0.908377  0.163097 -0.913962 -0.914933 -0.372983   
117053   74524.0  0.908828 -1.404905  1.314052  0.959853 -1.777457  0.705762   
159299  112396.0  2.025331 -0.126615 -1.419791  0.079005  0.209970 -0.558672   
...          ...       ...       ...       ...       ...       ...       ...   
178208  123525.0 -5.904921  4.439911 -8.631802  7.788684 -4.989580 -1.200144   
181966  125200.0 -0.769172  1.342212 -2.171454 -0.151513 -0.648374 -0.973504   
182992  125612.0  1.889618  1.073099 -1.678018  4.173268  1.015516 -0.009389   
183106  125658.0  0.224414  2.994499 -3.432458  3.986519  3.760233  0.165640   
184379  126219.0 -1.141559  1.927650 -3.905356 -0.073943 -0.044858 -1.756999   

              V7        V8        V9  .

In [25]:
print(Y)

168573    0.0
168292    0.0
173752    0.0
117053    0.0
159299    0.0
         ... 
178208    1.0
181966    1.0
182992    1.0
183106    1.0
184379    1.0
Name: Class, Length: 736, dtype: float64


**Splitting the dataset into testing and training**

In [27]:
#Splitting the dataset into testing and training
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,stratify=Y,random_state=2)

In [28]:
print(X.shape,X_train.shape,X_test.shape)

(736, 30) (588, 30) (148, 30)


**Model Training**

**Logistic Regression**

In [29]:
model=LogisticRegression()

**Training the model with logistic regression on the training data**

In [31]:
model.fit(X_train, Y_train)

LogisticRegression()

**Model Evaluation :Measuring the accurracy of the model**

In [52]:
joblib.dump(model,'credit-card-fraud-detection.joblib')

['credit-card-fraud-detection.joblib']

In [32]:
#Measuring the accurracy of the model using accuracy_score on the training dataset
X_train_prediction=model.predict(X_train)
training_data_accuracy=accuracy_score(X_train_prediction,Y_train)


In [35]:
print('Accuary of the training dataset is : ',training_data_accuracy)

Accuary of the training dataset is :  0.9081632653061225


**Thus we can see our accuracy score is 91%(apporx.) that means the prediction will be around 91% accurate so thus it concludes that our sampling of the dataset is also fine**

**Now, measuring the accurracy of the model using accuracy_score on the testing dataset**

In [34]:
#Measuring the accurracy of the model using accuracy_score on the testing dataset
X_test_prediction=model.predict(X_test)
testing_data_accuracy=accuracy_score(X_test_prediction,Y_test)

In [36]:
print('Accuary of the testing dataset is : ',testing_data_accuracy)

Accuary of the testing dataset is :  0.9391891891891891


**Thus our testing dataset is also having good accuracy score of 94% so predicting would be more accurate**

In [65]:
predicting_model=joblib.load('/content/credit-card-fraud-detection.joblib')
predictions=predicting_model.predict([[167543,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234,-0.234]])
print(predictions)

[0.]


/usr/local/lib/python3.8/dist-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


**Thus you can see that the model has now predicting when i entered cretain values ,and the prediction was that the particular transaction is not fraud or can say its *LEGIT(0)*** 